table/block_based_table_builder.cc

```diff
@@ -101,83 +101,123 @@ bool GoodCompressionRatio(size_t compressed_size, size_t raw_size) {
+          raw.data(),
+          raw.size(),
+          compressed_output);
     case kXpressCompression:
-      if (XPRESS_Compress(raw.data(), raw.size(),
-          compressed_output) &&
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;
+      return XPRESS_Compress(raw.data(), raw.size(), compressed_output);
     case kZSTD:
     case kZSTDNotFinalCompression:
-      if (ZSTD_Compress(compression_ctx, raw.data(), raw.size(),
-                        compressed_output) &&
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;     // fall back to no compression.
-    default: {}  // Do not recognize this compression type
+      return ZSTD_Compress(
+          compression_ctx, raw.data(), raw.size(), compressed_output);
+    default:
+      // Do not recognize this compression type
+      return false;
   }
+}
+
+} // namespace
 
-  // Compression method is not supported, or not good compression ratio, so just
-  // fall back to uncompressed form.
+// format_version is the block format as defined in include/rocksdb/table.h
+Slice CompressBlock(
+    const Slice& raw,
+    const CompressionContext& compression_ctx,
+    CompressionType* type,
+    uint32_t format_version,
+    std::string* compressed_output,
+    std::string* sampled_output_fast,
+    std::string* sampled_output_slow) {
+  *type = compression_ctx.type();
+
+  if (compression_ctx.type() == kNoCompression &&
+      !compression_ctx.sample_for_compression()) {
+    return raw;
+  }
+
+  // If requested, we sample one in every N block with a
+  // fast and slow compression algorithm and report the stats.
+  // The users can use these stats to decide if it is worthwhile
+  // enabling compression and they also get a hint about which
+  // compression algorithm wil be beneficial.
+  // FIXME-SD: replace random()
+  if (compression_ctx.sample_for_compression() &&
+      ((random() % compression_ctx.sample_for_compression()) == 0) &&
+      sampled_output_fast && sampled_output_slow) {
+    // Sampling with a fast compression algorithm
+    if (LZ4_Supported() || Snappy_Supported()) {
+      CompressionType c = LZ4_Supported() ? kLZ4Compression : kSnappyCompression;
+      CompressionContext ctx(c, compression_ctx.sample_for_compression(),
+                             compression_ctx.options());
+      CompressBlockInternal(raw,
+                            ctx,
+                            c,
+                            format_version,
+                            sampled_output_fast);
+    }
+
+    // Sampling with a slow but high-compression algorithm
+    if (ZSTD_Supported() || Zlib_Supported()) {
+      CompressionType c = ZSTD_Supported() ? kZSTD : kZlibCompression;
+      CompressionContext ctx(c, compression_ctx.sample_for_compression(),
+                             compression_ctx.options());
```

Comment:
I'm not sure CompressionContext is where the sample rate should go to. @ajkr will be the best person to determine.

Reply:
I debated not passing it at all and just choosing a fixed sampling rate internally but I thought it may be better to have an option there. Will talk to @ajkr .

Reply:
I think putting it inside `CompressionContext` is ok for now. The thing that looks a bit confusing is we have `CompressionContext` with file lifetime (held by `Rep`), and also another `CompressionContext` is created for each call to `CompressBlock`.

After #4849, there'll be a separate structure for per-`CompressBlock` data, i.e., `CompressionInfo`. I can move your sample rate there later.

Reply:
Well, it looks like `CompressionInfo` will need to be a per-`CompressBlockInternal` struct to work with this PR. Unless it is possible for you to setup the sample rate in the caller to `CompressBlock`? If you want sample rate to only apply to data blocks, `BlockBasedTableBuilder::WriteBlock` might be a good place for it. (Although perhaps I'm just biased towards making merge conflicts easier :)).